contracts/vaults/VaultPauser/RibbonVaultPauser.sol

```diff
@@ -0,0 +1,253 @@
+// SPDX-License-Identifier: MIT
+pragma solidity =0.8.4;
+import {SafeMath} from "@openzeppelin/contracts/utils/math/SafeMath.sol";
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+
+import {
+    SafeERC20
+} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import {
+    ReentrancyGuardUpgradeable
+} from "@openzeppelin/contracts-upgradeable/security/ReentrancyGuardUpgradeable.sol";
+import {
+    OwnableUpgradeable
+} from "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import {
+    ERC20Upgradeable
+} from "@openzeppelin/contracts-upgradeable/token/ERC20/ERC20Upgradeable.sol";
+import {IVaultPauser} from "../../interfaces/IVaultPauser.sol";
+import {Vault} from "../../libraries/Vault.sol";
+import {IRibbonThetaVault} from "../../interfaces/IRibbonThetaVault.sol";
+import {IWETH} from "../../interfaces/IWETH.sol";
+import {RibbonVault} from "../BaseVaults/base/RibbonVault.sol";
+import {ShareMath} from "../../libraries/ShareMath.sol";
+import {
+    RibbonVaultPauserStorage
+} from "../../storage/RibbonVaultPauserStorage.sol";
+
+contract RibbonVaultPauser is
+    ReentrancyGuardUpgradeable,
+    OwnableUpgradeable,
+    IVaultPauser,
+    RibbonVaultPauserStorage
+{
+    using SafeERC20 for IERC20;
+    using SafeMath for uint256;
+
+    /************************************************
+     *  NON UPGRADEABLE STORAGE
+     ***********************************************/
+
+    /// @notice Stores all the vault's paused positions
+    struct PauseReceipt {
+        uint16 round;
+        address account;
+        uint128 shares;
+    }
+
+    mapping(address => mapping(address => PauseReceipt[]))
+        public pausedPositions;
```

Comment:
For simplicity, could we just do `address => PauseReceipt`? So we make it so that one user can only have one pause per vault. ie they can only pause their entire balance